signbank.py

```diff
@@ -0,0 +1,331 @@
+import csv
+import os
+import re
+import shutil
+import sqlite3
+import sys
+
+import requests
+from requests.adapters import HTTPAdapter, Retry
+
+DEFAULT_SIGNBANK_HOST = os.getenv("SIGNBANK_HOST", "https://signbank.nzsl.nz")
+SIGNBANK_DATASET_ID = os.getenv("SIGNBANK_DATASET_ID", 1)
+SIGNBANK_USERNAME = os.getenv("SIGNBANK_USERNAME")
+SIGNBANK_PASSWORD = os.getenv("SIGNBANK_PASSWORD")
+
+##
+# Start a requests session that is authenticated to Signbank
+
+
+def signbank_session():
+    s = requests.Session()
+    s.get("%s/accounts/login/" % DEFAULT_SIGNBANK_HOST)
+    s.post("%s/accounts/login/" % DEFAULT_SIGNBANK_HOST,
+           data={'username': SIGNBANK_USERNAME, 'password': SIGNBANK_PASSWORD,
+                 'csrfmiddlewaretoken': s.cookies['csrftoken']},
+           headers={'Referer': DEFAULT_SIGNBANK_HOST})
+
+    return s
+
+##
+# Start a requests session that is set up to retry S3 requests.
+# This resolves an issue where S3 occasionally resets the connection
+# when downloading a large number of files. It also allows for temporary
+# failures in a request.
+
+
+def s3_session():
+    s = requests.Session()
+    retries = Retry(total=5, backoff_factor=1,
+                    status_forcelist=[502, 503, 504])
+    s.mount('https://', HTTPAdapter(max_retries=retries))
+
+    return s
+
+##########################
+# Dictionary data handling
+##########################
+
+
+def fetch_gloss_export_file(filename):
+    session = signbank_session()
+    response = session.get("%s/dictionary/advanced/" % DEFAULT_SIGNBANK_HOST,
+                           params={"dataset": SIGNBANK_DATASET_ID, "published": 'on', "format": 'CSV'})
+    response.raise_for_status()
+    with open(filename, "wb") as f:
+        f.write(response.content)
+
+
+def parse_signbank_csv(filename):
+    with open(filename, 'r') as f:
+        reader = csv.reader(f)
+        headers = next(reader, None)
+        return [{h: x for (h, x) in zip(headers, row)} for row in reader]
+
+
+##########################
+# Asset handling
+##########################
+def fetch_gloss_asset_export_file(filename):
+    session = signbank_session()
+    video_response = session.get("%s/video/csv" % DEFAULT_SIGNBANK_HOST)
+    video_response.raise_for_status()
+    with open(filename, "wb") as f:
+        f.write(video_response.content)
+
+
+def fetch_gloss_assets(data, database_filename, output_folder):
+    if not os.path.exists(output_folder):
+        os.makedirs(output_folder)
+
+    db = sqlite3.connect(database_filename)
+    db.executescript(
+        """
+        BEGIN;
+        CREATE TABLE videos (
+            word_id, video_type, filename, url, display_order
+        );
+        CREATE UNIQUE INDEX idx_word_videos ON videos (word_id, video_type, filename);
+        COMMIT;
+      """
+    )
+
+    for entry in data:
```

Comment:
I wonder if this would be easier to read if it was broken up into smaller methods?

Reply:
Yeah, I'm refactoring this into an image processing module so we can share this between Freelex and Signbank

Reply:
@mischa-s https://github.com/ODNZSL/nzsl-dictionary-scripts/pull/12/commits/940ebb18dbb06760f00ff361e88d0691307911a5

Reply:
@joshmcarthur that commit looks good but doesn't appear to be used in this file, are you saying you will do that as a next step or have I misunderstood?

Reply:
Ah, I thought this was the image processing images. I'm not sure if we gain much by splitting this one up. Python methods tend to run longer anyway in my experience, but in this case there's quite a lot of dependency on checking video types, filenames etc across the 3 inserts.

Reply:
Okay happy to 🐘 this - I think there's still probably some refactoring wins, but not too fussed